apps/web/src/pages/settings/ApiKeysPage/ApiKeysPage.tsx

```diff
@@ -0,0 +1,109 @@
+
+  return (
+    <>
+      <Flex direction={'column'} gap={'200'} maxW="37.5rem">
+        <Input
+          readOnly
+          type={isApiKeyMasked ? 'password' : 'text'}
+          label="API Key"
+          description="Use this API key to interact with the Novu API"
+          rightSection={
+            // this is ugly, but we define the width of rightSection explicitly, which messes with larger elements
+            <Flex gap="125" position={'absolute'} right="100">
+              <IconButton
+                onClick={openModal}
+                tooltipProps={{ label: 'Regenerate API Key' }}
+                id={'api-key-regenerate-btn'}
+              >
+                <IconRefresh size={INPUT_ICON_SIZE} />
+              </IconButton>
+              <IconButton
+                onClick={toggleApiKeyVisibility}
+                tooltipProps={{ label: isApiKeyMasked ? 'Show API Key' : 'Hide API Key' }}
+                id={'api-key-toggle-visibility-btn'}
+              >
+                {isApiKeyMasked ? (
+                  <IconOutlineVisibility size={INPUT_ICON_SIZE} />
+                ) : (
+                  <IconOutlineVisibilityOff size={INPUT_ICON_SIZE} />
+                )}
+              </IconButton>
+              <ClipboardIconButton
+                isCopied={clipboardApiKey.copied}
+                handleCopy={() => clipboardApiKey.copy(apiKey)}
+                testId={'api-key-copy'}
+                size={INPUT_ICON_SIZE}
+              />
+            </Flex>
+          }
+          value={apiKey}
+          data-test-id="api-key"
+        />
+        <Input
+          readOnly
+          label="Application Identifier"
+          description="A public key identifier that can be exposed to the client applications"
+          rightSection={
+            <ClipboardIconButton
+              isCopied={clipboardEnvironmentIdentifier.copied}
+              handleCopy={() => clipboardEnvironmentIdentifier.copy(environmentIdentifier)}
+              testId={'application-identifier-copy'}
+              size={INPUT_ICON_SIZE}
+            />
+          }
+          value={environmentIdentifier}
+          data-test-id="application-identifier"
+        />
+        <Input
+          readOnly
+          label="Environment ID"
+          rightSection={
+            <ClipboardIconButton
+              isCopied={clipboardEnvironmentId.copied}
+              handleCopy={() => clipboardEnvironmentId.copy(environmentId)}
+              testId={'environment-id-copy'}
+              size={INPUT_ICON_SIZE}
+            />
+          }
+          value={environmentId}
+          data-test-id="environment-id"
+        />
+      </Flex>
+      <ConfirmRegenerationModal {...modalComponentProps} />
+    </>
+  );
+};
+
+export const ApiKeysPage = () => {
+  const props = useApiKeysPage();
+
+  return (
+    <SettingsPageContainer title={`API keys (${props.pageEnv})`}>
+      <ApiKeysRenderer {...props} />
+    </SettingsPageContainer>
```

Comment:
I'm not sure about this pattern, like I see the value when we have something that could be reusable, but IMO we create more hassle with props drilling, `ApiKeysPage -> ApiKeysRenderer -> ConfirmRegenerationModal`. Just the my opinion, don't take it as a must.

Reply:
I can certainly understand that! There are two things that I'd add to considerations here:

1. **General**: I agree we are not likely to reuse the whole page content. However, this is not only for reusability, but also makes things much easier to test when doing integration tests (since we can pass whatever we want as props and validate against that).
2. **Specific**: While this `ConfirmRegenerationModal` may not be particularly applicable to other situations, I think our implementation strategy for modals is currently broken, and this case shows a major part of it. Our modals should not have their "roots" placed on each page IMO. Instead, we could have a `ModalProvider` that is included in the base layout and register modals by a key. Then we can call: `showModal(ModalKey.REGENERATE_API_KEYS)`. More realistically, we'd see that most modals fall into a few main varieties -- like confirmation (yes / no), acknowledgement (ok), etc. Then we can more simply do `showModal(ModalKey.CONFIRMATION, { /* props */ })` rather than writing one-off components.